### PR TITLE
Add Subscriptions to Activation Keys by Upstream Pool ID

### PR DESCRIPTION
**Pull request by wbclark**

### CODE REVIEW

File: plugins/modules/activation_key.py [hunk 1]
```diff
@@ -275,10 +283,14 @@ def main():
             if subscriptions is not None:
                 desired_subscriptions = []
                 for subscription in subscriptions:
-                    if subscription.get('name') is not None and subscription.get('pool_id') is None:
+                    if subscription.get('name') is not None and subscription.get('pool_id') is None and subscription.get('upstream_pool_id') is None:
```
Comment (mdellweg) [inline plugins/modules/activation_key.py]: If the required_one_of and mutually_exclusive work properly, this logic can be simplified a lot.

Comment (evgeni) [reply plugins/modules/activation_key.py]: that's true, I think this check pre-dates the `required_one_of` and `mutually_exclusive` above, worth checking if we can drop it

File: plugins/modules/activation_key.py [hunk 2]
```diff
@@ -47,21 +47,28 @@
...
           - Pool ID of the Subscription to be added.
-          - Mutually exclusive with I(name).
+          - Mutually exclusive with I(name) and I(upstream_pool_id).
+        type: str
+        required: false
+      upstream_pool_id:
+        description:
+          - Upstream Pool ID of the Subscription to be added.
+          - Mutually exclusive with I(name) and I(pool_id).
+          - Supported since Katello 3.15.0
```
Comment (evgeni) [inline plugins/modules/activation_key.py]: I must admit I prefer not to have version numbers in doc strings, for two reasons:
* upstream moves quickly (heck, 3.15 is not supported anymore ;)) and this makes the comment kinda out of date quickly
* downstream the users don't necessarily know which component versions they have.

I wonder if there is a trivial way to deduct support for that param from the apidoc and warn the user in the case it's not supported.

either way, shouldn't block this pr

**Comment by evgeni** [comment]
can you please add a changelog fragment in `changelogs/fragments`? something like:

```yaml
minor_changes:
  - activation_key - add support for selecting subscriptions by ``upstream_pool_id``
```

**Comment by wbclark** [comment]
#541 is worded in a way that is pretty general but we could consider closing it with this

**Comment by wbclark** [comment]
@mdellweg @evgeni Thanks for the reviews. I've pushed an update based on your feedback.

File: plugins/modules/activation_key.py [hunk 3]
```diff
@@ -275,10 +282,14 @@ def main():
...
                 for subscription in subscriptions:
-                    if subscription.get('name') is not None and subscription.get('pool_id') is None:
+                    if subscription.get('name') is not None:
                         desired_subscriptions.append(module.find_resource_by_name('subscriptions', subscription['name'], params=scope, thin=True))
                     if subscription.get('pool_id') is not None:
                         desired_subscriptions.append(module.find_resource_by_id('subscriptions', subscription['pool_id'], params=scope, thin=True))
+                    if subscription.get('upstream_pool_id') is not None:
+                        desired_subscriptions.append(
+                            module.find_resource_by('subscriptions', 'upstream_pool_id', subscription['upstream_pool_id'], params=scope, thin=True)
+                        )
```
Comment (mdellweg) [inline plugins/modules/activation_key.py:292]: :+1:

**Review by mdellweg — APPROVED** [review]

**Comment by evgeni** [comment]
Is upstream pool id unique? Or do we have the same issue as with the name that we get multiple subs that have the same name but different expiration dates or support levels?

I'd not close #541 here

**Review by evgeni — APPROVED** [review]

**Comment by mdellweg** [comment]
> Is upstream pool id unique? Or do we have the same issue as with the name that we get multiple subs that have the same name but different expiration dates or support levels?

The module Docs (in this PR) state that they are not unique.

**Comment by evgeni** [comment]
Then let's not close #541, thanks!

**Comment by wbclark** [comment]
@evgeni @mdellweg 

To answer the previous question, Upstream Pool IDs are "more unique" than Subscription Names, but not guaranteed to be fully unique.

It used to be the case that entitlement count for an existing pool on a manifest could not be modified on the RH portal, so an user wishing to add additional entitlements from the same upstream pool could only create a new downstream pool with the additional entitlements.

Now the 'subscription allocations' interface on the RH portal is updated and allows modifying the account of existing downstream pools and does not allow creating a separate downstream pool from the same upstream pool; however, nothing prevents that from existing on a manifest and users could still use an older manifest that was created in that fashion.

What is worse is that additionally there is an interface in Katello itself for adding subscriptions to the manifest; this creates a 'bind entitlements to allocation' task that, when additional entitlements are added from an upstream pool which is already used on the manifest, does in fact create a separate downstream pool instead of updating the quantity on the existing downstream pool.

Interestingly, for subscriptions that create a derived entitlement that is for 'temporarily unmapped guests', the Candlepin database stores the upstream pool IDs for these meaning they are not unique for Candlepin itself, but the upstream_pool_id field in the katello_pools table of Foreman database is actually blank for those derived entitlements. (And there are edge cases where an user would want to attach such derived entitlements to an activation key. To my knowledge that would happen with a 'guests of virt-who-*' entitlement and not with the 'temporarily unmapped guest' entitlement, but that still presents a barrier to being able to use upstream pool IDs in all cases).

After researching and confirming these issues, I agree with the call to leave #541 open.

**Comment by evgeni** [comment]
Whoa, thanks for that elaborate explanation! Much appreciated!

**Pull request merged**